src/Main.elm

```diff
@@ -1,10 +1,829 @@
+        , E.height <| E.px (round width)
+        ]
+        <| E.el
+            [ EBackground.color color_
+            , E.width <| E.px w
+            , E.height <| E.px w
+            , E.centerX
+            , E.centerY
+            , E.htmlAttribute <| Html.Attributes.style "border-radius" "50%"
+            ]
+            <| E.none
+
+
+
+{-
+   Note on error reporting:
+   When I saw these ids that represent "parents" or "folders" I thought a lot
+   could go wrong.
+   What if we have a folder whose parent id actually points to an "audience"? Or
+   an audience whose "folder" id actually points to another "audience"?
+   What if there there's cyclical references? folder 1 points to folder 2 and vice versa?
+   What if we have a "parent" or "folder" id that points to a folder or parent that
+   doesn't exist?
+
+   Initially I thought "ok, cool maybe we can do something in elm to address this".
+   But then I realized that that would probably make the app much slower. Having to
+   to check for all of these potential errors would bring a lot of complexity
+   code-wise and also time-complexity wise. So instead I concluded that this may be
+   better addressed on the backend by writing very good tests.
+   This way our frontend is faster, easier to understand and our tests for these issues
+   only live in one place: the backend.
+
+   Additionally, this made me think that since there are so many things that could go
+   wrong by having this model of references between the audiences and audienceFolders,
+   maybe this would even better be addressed not by writing really good tests on the
+   backend, but instead by data modelling and representing these relationships in a
+   better way, thus making these invalid states impossible to even represent, so we
+   wouldn't even need to test for them.
+
+   Relevant: Richard Feldman's talk "Making Impossible States Impossible"
+   https://www.youtube.com/watch?v=IcgmSRJHu_8
+
+   Just for the record: In production code I wouldn't write these sorts of very long
+   comments. I'd instead probably open up an issue or ping the team on slack.
+-}
+
+-- I haven't spent much time styling this because this is something the user shouldn't
+-- even see. No point in styling it just for it to look cute for the developers
+viewJsonError : Json.Decode.Error -> E.Element Msg
+viewJsonError err =
+    let
+        formattedErrText =
+            Json.Decode.errorToString err
+                |> String.replace "\\n" "\n"
+                |> String.replace "\\\"" "\""
+    in
+    E.el
+        [ EBackground.color <| E.rgb255 240 80 105
+        , E.width E.fill
+        , E.paddingXY 20 20
+        , EBorder.rounded 7
+        ]
+        <|
+            E.el
+                [ E.centerX
+                , E.centerY
+                ]
+                <| E.text formattedErrText
+
+
+
+-- intermediate representation for "audience"s and "audienceFolder"s , so we
+-- can easily make dicts out of them
```

Comment:
What's wrong about usage Audience directly and have it converted to dict this way
```Elm
audiences
        |> List.map (\a -> ( a.id, a ))
        |> Dict.fromList
```
It's more beneficial for future changes in data structure. If Audience will be extended let's say by `createdAt : Posix` you need to change this on two places instead of one. It's not big chunk of work, but what if you will have more modules where `Data/Audiences` are used. It will require at least share `FormattedAudience` too.

Reply:
Yeah, you're right. My thought was that it was nicer because since we're working with dicts, it'd be a bit clearer to have a function name there that literally says "audiencesToDict" or something like that, but now that you mention it, I agree that things would go out of sync, and it'd just make you unnecessarily change code in 2 places, just for the id field which is different.